sgv2-dynamo-api/src/main/java/io/stargate/sgv2/dynamosvc/parser/FilterExpressionVisitor.java

```diff
@@ -0,0 +1,148 @@
+
+  @Override
+  public Object visitAndExpr(ExpressionParser.AndExprContext ctx) {
+    if (Boolean.FALSE.equals(this.visit(ctx.left))) {
+      return Boolean.FALSE;
+    }
+    return this.visit(ctx.right);
+  }
+
+  @Override
+  public Object visitOrExpr(ExpressionParser.OrExprContext ctx) {
+    if (Boolean.TRUE.equals(this.visit(ctx.left))) {
+      return Boolean.TRUE;
+    }
+    return this.visit(ctx.right);
+  }
+
+  @Override
+  public Object visitTerm(ExpressionParser.TermContext ctx) {
+    return this.visit(ctx.children.get(0));
+  }
+
+  @Override
+  public Object visitOpTerm(ExpressionParser.OpTermContext ctx) {
+    AttributeValue realValue = (AttributeValue) this.visit(ctx.name());
+    if (realValue == null) {
+      // current item does not have this attribute
+      return false;
+    }
+    AttributeValue compareValue = (AttributeValue) this.visit(ctx.value());
+    String operator = ctx.operator.getText();
+    switch (operator) {
+      case "=":
+        return Objects.equals(realValue, compareValue);
+      case "<>":
+        return !Objects.equals(realValue, compareValue);
+    }
+    switch (operator) {
+      case ">":
+        return DataMapper.compareTo(realValue, compareValue) > 0;
+      case ">=":
+        return DataMapper.compareTo(realValue, compareValue) >= 0;
+      case "<":
+        return DataMapper.compareTo(realValue, compareValue) < 0;
+      case "<=":
+        return DataMapper.compareTo(realValue, compareValue) <= 0;
+      default:
+        throw new IllegalArgumentException("Operator " + operator + " is invalid");
+    }
+  }
+
+  @Override
+  public Object visitBetweenTerm(ExpressionParser.BetweenTermContext ctx) {
+    AttributeValue lowerBound = (AttributeValue) this.visit(ctx.lower);
+    AttributeValue upperBound = (AttributeValue) this.visit(ctx.upper);
+    AttributeValue realValue = (AttributeValue) this.visit(ctx.name());
+    if (realValue == null) {
+      // current item does not have this attribute
+      return false;
+    }
+    return DataMapper.compareTo(realValue, lowerBound) >= 0
+        && DataMapper.compareTo(realValue, upperBound) <= 0;
+  }
+
+  @Override
+  public Object visitBeginTerm(ExpressionParser.BeginTermContext ctx) {
+    AttributeValue realValue = (AttributeValue) this.visit(ctx.name());
+    AttributeValue beginValue = (AttributeValue) this.visit(ctx.value());
+    if (realValue == null) {
+      // current item does not have this attribute
+      return false;
+    }
+    Preconditions.checkArgument(
+        realValue.getS() != null && beginValue.getS() != null,
+        "begins_with condition only applies to string type");
+    return realValue.getS().startsWith(beginValue.getS());
+  }
+
+  /**
+   * Given an attribute's name (could be top-level attribute or a nested attribute), retrieve its
+   * corresponding value provided by database
+   *
+   * @param ctx
+   * @return
+   */
+  @Override
+  public Object visitName(ExpressionParser.NameContext ctx) {
```

Comment:
This method looks very complicated (and possibly slow) for what would sound like a simple value lookup.
And why is a lexer needed? Is it to handle dotted names ("root.branch.leaf") or something?

Reply:
Yes, lexer and parser are needed to handle dotted names and array offsets, e.g. `root.branch[0].leaf[1]`.